internal/internal_event_handlers.go

```diff
@@ -1003,3 +1028,13 @@ func (weh *workflowExecutionEventHandlerImpl) handleSignalExternalWorkflowExecut
 
 	return nil
 }
+
+// getCancellationID, helper function determine the ID used for workflow cancellation
+// this function exists for backward compatibility, i.e. prev generation of client use workflow ID as identifier,
+// and there is no control (being empty / missing), so we have to use the workflow ID
+func getCancellationID(control []byte, workflowID string) string {
```

Comment:
this might cause problem. for example, replay the history that was generated by the old version code, so the control will be empty from history, but the state machine is keyed with the cancellation id.

Reply:
existing client use workflow ID for indexing.
new client use control for indexing.

existing client will always use workflow ID for indexing, because existing client do not use control at all. so existing client should not break.

new client will work if replay using events from a new client, since both are using control for indexing.

new client will still work if replay using events from a old client, because old client will not touch the control, leaving that empty, so new client will use workflow ID for indexing, same as old client.

Reply:
one more thing, control will only be set by new client, and will be copy pasted by service
so as long as control is set on the decision (by new client), it will appear on all related events, if not (by the old client), it will not appear.

Reply:
ok there might be a issue, let me check

Reply:
i think it will be ok if in the reply mode, client initialize 2 cancellation state machine, one using workflow ID, one using cancellation ID. once the Cancellation Initiated Event is received, one of the above 2 state machine can be removed.